### PR TITLE
New module for the tool qc3C.

### PR DESCRIPTION
**Pull request by cerebis**
This is a new module for support of a new Hi-C quality control tool called qc3C.

 - [x] This comment contains a description of changes (with reason)
 - [x] `CHANGELOG.md` has been updated
 - [x] There is example tool output for tools in the https://github.com/ewels/MultiQC_TestData repository or attached to this PR
 - [x] Code is tested and works locally (including with `--lint` flag)
 - [x] `docs/README.md` is updated with link to below
 - [x] `docs/modulename.md` is created
 - [x] Everything that can be represented with a plot instead of a table is a plot
 - [x] Report sections have a description and help text (with `self.add_section`)
 - [x] There aren't any huge tables with > 6 columns (explain reasoning if so)

There is a wider table which contains simple runtime options and alternating column colours. The reinterpretable results are all much simpler or plots.

 - [x] Each table column has a different colour scale to its neighbour, which relates to the data (eg. if high numbers are bad, they're red)
 - [x] Module does not do any significant computational work

### CODE REVIEW

**Comment by cerebis** [comment]
These tests should pass now that the up-to-date test data has been commited to MultiQC_TestData

**Comment by cerebis** [comment]
Though I would like to say otherwise, it's possible I will break my own JSON file in the future. I have added a large try/except block over the parsing method, which will stop gracefully after logging an error.

File: multiqc/modules/qc3C/qc3C.py [hunk 1]
```diff
@@ -0,0 +1,762 @@
...
+            self.digest_junctions[analysis_mode][s_name] = \
+                [{'name': juncs[i][1].decode(), 'color': cols[i]} for i in range(len(juncs))]
+
+            self.add_data_source(f, s_name, section=analysis_mode)
+
+        except KeyError as ex:
+            import os
+            log.error('An error ocurred while parsing qc3C JSON file {}'
+                      .format(os.path.join(f['root'], f['fn'])))
+            raise UserWarning(ex)
```
[the remark this answers is not in the document]
Comment (cerebis) [reply multiqc/modules/qc3C/qc3C.py:762]: Would you agree to just log.exception(ex) and not raise the error?

[the remark this answers is not in the document]
Comment (ewels) [reply multiqc/modules/qc3C/qc3C.py:762]: Untested, but I think something like this should work:

```suggestion
        except KeyError as ex:
            log.error('Missing expected key '{}' for file: {}'.format(ex.args[0], os.path.join(f['root'], f['fn'])))
```

[the remark this answers is not in the document]
Comment (ewels) [reply multiqc/modules/qc3C/qc3C.py:762]: (So basically, yes 👍 )

**Comment by ewels** [comment]
If it's ok with you, I will quickly have a stab at editing this code myself now. This is mostly to change the whitespace / formatting conventions to "MultiQC style" (eg. my personal style). This might be a little annoying and I apologise, but I find that keeping a consistent code style across all MultiQC modules really helps.

At the same time I'll clear up and minor bugs / tweaks that I see as I go. Then I'll come back with some review comments on anything bigger.

**Comment by ewels** [comment]
By the way, this is super cool! We were discussing writing a tool exactly along these lines at SciLifeLab / NGI with @remiolsen, so we're really happy that you beat us to it! 😅

**Comment by cerebis** [comment]
Sure, feel free.

This code isn't completely representative of my own predilections either.
Such as the length of the parse_qc3c_log() method and how I am passing the
data around.

I will stop making commits.

**Comment by ewels** [comment]
To be clear - these changes are absolutely not in criticism of your coding style. This stuff is highly personal and I am likely making things arguably worse in places / being hypocritical. I'm just trying to save my future-self (and others) a little mental work by keeping stuff consistent where possible.

**Comment by cerebis** [comment]
Ah great. I have spent far too long on this idea and originally intended to
publish an article on it  _last_ year. After the time spent, I really
wanted a nice report and didn't have the time or desire to implement it.

So, it was great that I was introduced to MultiQC and appreciated how it
gathered results and the overall appearance. Then other day I thought, well
do these guys really implement parsing these tools themselves? Glad to see
you had the whole process outlined.

**Comment by ewels** [comment]
Ok, my annoying reformatting stuff is done - sorry again about that.

Aside from whitespace, the main other thing I did was move a lot of the section `description` texts into the `helptext`. This makes it hidden by default in the report but easily findable for anyone looking for longer explanations:

Jun-16-2020 17-16-29

I also prefixed many of the HTML anchors and IDs with `qc3C_`, to make them a little more specific to this module. This apparently broke the linting somehow as we're now getting warnings about identical table column IDs. I'll take a look at that later, not a big deal right now.

### Suggested changes

From looking at the report, I have a few general suggestions:

* Some table columns can be removed
    * Only the really key stats should make it into MultiQC reports. Stuff like `Date`, `Run Mode`, `Seed` etc should be removed (especially as this table only shows when `Run Mode` is `bam` :wink:)
    * Some columns should maybe be hidden by default as a compromise. For example, where values are typically the same for all samples (but I prefer stuff to just be removed where possible).
* Too many tables
    * Some of the tables could probably be merged, especially if we drop some columns.
    * The really key / more general columns could be moved to the special "General Statistics" MultiQC table so that they show up alongside metrics from other tools.
    * I wonder if the `read-pair breakdown` tables could be a couple of bar plots instead? I'm not totally sure how / if the percentages could stack though.
    * The `BAM mode HiCPro categories` seems an exact duplicate of the bar graph below
    * The `BAM mode long-range pairs` table is very similar to the line graph underneath it
        * If you really want to keep this, then it could could be a bar graph instead.
* Some plots may be better if they are conditional
    * I'm not 100% sure I'm interpreting this correctly, but I think that the `BAM mode junction breakdown` could have just one category if all samples are single-enzyme? So maybe we could check for that and skip that section if it's the case.
* Order of categories
    * Bar graph categories should typically be Good -> Bad. `* mode read parsing` and `K-mer mode junction proportion` categories should be flipped.

I don't love having so many sections which are very similar between bam mode and kmer mode, but I can't see an easy way around that. I guess that most people will only run their samples in one of the two modes anyway? So maybe not an issue.

The general gist is that I dislike tables, and I like to keep report content as slim and easy to read as possible.

### Small stuff:

* Line graphs have [built-in functionality](https://multiqc.info/docs/#line-graphs) for log x-axis, no need to calculate this yourself 
    * This is better than calculating those values yourself, as the plot axis then displays the raw counts instead of the log value.
* `K-mer mode junction proportion` plot subsampling
    * If this plot always uses a subsampled read count and the "Number of reads" part of the plot is not super informative, you could make this percentage only. `'cpswitch_c_active': False` + `'cpswitch': False` should do this I think.
* Any read counts in tables should use `'shared_key': 'read_count'` and the associated `read_count_multiplier` stuff - see [docs](https://multiqc.info/docs/#step-3-adding-to-the-general-statistics-table).
   * Could be good to set `shared_key` elsewhere in tables too. Then a consistent scale for the background bar will be used across those columns - good for example with insert lengths etc.

Ok, I need to run now, but please let me know what you think about the above. From my run over the code I didn't really see any issues so I think it's pretty much just this more conceptual stuff.

Looking good! Thanks again for working on this!

Phil

**Comment by cerebis** [comment]
> Aside from whitespace, the main other thing I did was move a lot of the section `description` texts into the `helptext`. This makes it hidden by default in the report but easily findable for anyone looking for longer explanations:

Thanks. I'd considered using helptext and comment but was not confident that users would actually think to click on the button. If its the normal mode of use, I suppose everyone will be aware of the value.

> I also prefixed many of the HTML anchors and IDs with `qc3C_`, to make them a little more specific to this module. This apparently broke the linting somehow as we're now getting warnings about identical table column IDs. I'll take a look at that later, not a big deal right now.

> ### Suggested changes

> * Some table columns can be removed

I have hidden the more redundant or uninteresting columns, but would like to keep them as they are at least useful to me when copying out of MultiQC. 

qc3C has two primary modes of operation, which ultimately both produce a estimate for the fraction of Hi-C read-pairs. But, our understanding of how well these figures agree across a wide array of data quality and how to further interpret that difference is ongoing. 

Also, other details like seed and sample rate are important. Ideally,  small sample sizes are preferable from a "Pilot->QC->Full Run" workflow. Thus far, small samples appear to give consistent results, however I suspect users to bring me results over millions of read-pairs -- especially when a library has turned out to be of poor quality and they wish it to be otherwise. Having the possibility of it all at hand within MultiQC will be useful.

> * Too many tables
>   
>   * Some of the tables could probably be merged, especially if we drop some columns.

Would these not become too wide?

>   * The really key / more general columns could be moved to the special "General Statistics" MultiQC table so that they show up alongside metrics from other tools.
>   * I wonder if the `read-pair breakdown` tables could be a couple of bar plots instead? I'm not totally sure how / if the percentages could stack though.

>   * The `BAM mode HiCPro categories` seems an exact duplicate of the bar graph below

This is true, I will disable the table.

>   * The `BAM mode long-range pairs` table is very similar to the line graph underneath it

This coarse cumulative histogram is akin to assembly statistics like N50 and something that is quoted within the literature. For some researchers, this fraction will be the gold-standard for assessing quality. I would really like it to be easily accessible in reports and not something that has to be moused-over to see details and then hand-transcribed.

It is possible copy out tables from graphs?

>     * If you really want to keep this, then it could could be a bar graph instead.
> * Some plots may be better if they are conditional
>   
>   * I'm not 100% sure I'm interpreting this correctly, but I think that the `BAM mode junction breakdown` could have just one category if all samples are single-enzyme? So maybe we could check for that and skip that section if it's the case.

The same figure is produced in either mode, but yes there is a single category for enzymes with non-ambiguous recognition sites. This type of digest was the norm not too long ago, but the commercial kits from both Phase and Arima use more complex digests. As the protocol is a challenge and these kits aim to improve success rate, its likely that the majority of libraries will possess multiple possible junctions.

That said, I can make it conditional.

> * Order of categories
>   
>   * Bar graph categories should typically be Good -> Bad. `* mode read parsing` and `K-mer mode junction proportion` categories should be flipped.

I have flipped the order for those bargraphs which had a Good->Bad context. These were the parsing results for eithe rmode and the k-mer junction proportion.

> 
> I don't love having so many sections which are very similar between bam mode and kmer mode, but I can't see an easy way around that. I guess that most people will only run their samples in one of the two modes anyway? So maybe not an issue.

Correct, this testing result won't be how people will really use the software. I think, however, it is important to distinguish between the source of the results in a manner that can't be overlooked by third-parties.


> ### Small stuff:
> * Line graphs have [built-in functionality](https://multiqc.info/docs/#line-graphs) for log x-axis, no need to calculate this yourself

I realise I could use a log axis, but dealing with a distribution which has an important feature at both the small end (where the peak exists) and for which the long-tail contains all the value is in Hi-C, I much prefer the bins to be log-scaled. This log-space binning is already a part of qc3C's output (whether displayed by MultiQC or not).

The x-axis is displaying as a TT with the correct values, while the y-axis is density. I add a 100x multiplier to it, so that the values don't go to zero. I would have liked `{:g}` to have worked, but Highcharts or JS do not seem to support that part of the printf mini-language.

> * `K-mer mode junction proportion` plot subsampling
>   
>   * If this plot always uses a subsampled read count and the "Number of reads" part of the plot is not super informative, you could make this percentage only. `'cpswitch_c_active': False` + `'cpswitch': False` should do this I think.

Agreed. 

At some in this work I had it that way, and should've returned it when I settled on a figure. It is debatable whether this figure has enough merit to be included, but it goes to the situation where a library has turned out very poor.  I have shown projects, which did not use pilots and had spent much on sequencing omly to get weak results. Occasionally there is push-back when low signal is inferred, and this very simple count helps to persuade.

> * Any read counts in tables should use `'shared_key': 'read_count'` and the associated `read_count_multiplier` stuff - see [docs](https://multiqc.info/docs/#step-3-adding-to-the-general-statistics-table).

>   * Could be good to set `shared_key` elsewhere in tables too. Then a consistent scale for the background bar will be used across those columns - good for example with insert lengths etc.

I wasn't quite sure how the shared_key worked. It was early into my experiments with MultiQC modules, but I did try it out and couldn't see the impact. 

What am I supposed to see as a visual effect? Also, is the key meant to refer to another header key or whether "shared_key" was yet another independent string that I am meant to use across many headers -- and if that was the case how does the logic work?

**Comment by ewels** [comment]
> Would these [tables] not become too wide?

Depends on how many columns are dropped / hidden 😄 

> It is possible copy out tables from graphs?

Not to copy, no - but the plot data can be downloaded as TSV from the toolbox. But I understand the analogy with the N50. If there's a good reason to have the table then we can keep it.

> I much prefer the bins to be log-scaled.

I don't completely understand what you've written here. I'm not suggesting that you use a linear scale, just that the logging is handled by highcharts so that the x-axis shows raw counts. See [this highcharts demo](https://www.highcharts.com/demo/line-log-axis) for an example. However, if the logged bins are calculated in qc3C and handed to MultiQC in that way then it will be a bit of a hassle to back convert, so I understand if you'd rather leave it as it is. My thinking at the time was just that it would be nice to have the numbers directly comparable to the table above.

> I wasn't quite sure how the shared_key worked.

Yeah, maybe the docs aren't super clear here (improvements to documentation always welcome! Also in the repo..). It basically just affects the bars in the table cell background. So if you have three columns showing read counts, eg. raw, trimmed, deduplicated, you could share the same x-axis across all three of those columns. So if the highest trimmed value is 50% of the highest raw value then it'll be 50% of the column width instead of 100%.

If the max values are very similar (eg. mean and median values) then you probably won't really see a visual difference.

The string can be anything, but it's only worth doing if you're going to share it across multiple columns in a single table. There are some conventions, mostly the `read_counts` label, which are useful for example in the General Statistics table where there will be columns from different modules which can then share their plotting x-axis.

I hope this makes sense!

Phil

**Comment by ewels** [comment]
ps. out of curiosity, how does the tool handle Omni-C data?

**Comment by cerebis** [comment]
> ps. out of curiosity, how does the tool handle Omni-C data?

Omni-C being the idea the using DNase or mechanical sheering to produce free-ends, rather than the use of restriction enzymes? In this case, I expect not.

qc3C exploits the predicability of junction sequences to make its assessments. If any two free-ends are possible, then qc3C as it is presently devised will not have a reference point -- so to speak.

**Comment by ewels** [comment]
Yup, as in the Dovetail kit: https://dovetailgenomics.com/omni-c/ Makes sense!

**Comment by cerebis** [comment]
> > I much prefer the bins to be log-scaled.
> 
> I don't completely understand what you've written here. I'm not suggesting that you use a linear scale, just that the logging is handled by highcharts so that the x-axis shows raw counts. 

I feel like we're misunderstanding each-other. For the sake of certainty, I am calling the horizontal axis X and vertical Y. The X axis is pair separation measured in base-pairs while the Y axis is density, rather than counts (Eg. scale_factor 1/N).

The demo is a good visual for me to try again to explain. I wish the bins over pair-separation to be uniformly spaced when on a log axis, so that both the peak near the origin is visible, as well has the long-range pairs being depicted. In the demo the bin widths vary, which makes it hard to effectively sample both something near the origin and the long tail without an excess of bins as separation grows.

However, I am re-reading what you said about x-axis and counts and wonder if you're looking at the plot with the x-axis along the vertical? If so, those are density values rather than counts. I need to normalise the curves if I wish them to appear over reasonably similar ranges.

**Comment by ewels** [comment]
oops, did you mean to close this PR @cerebis? I'll reopen as we're still working on it..

**Comment by ewels** [comment]
Regarding the log stuff - I do understand, you want logged pair separation bins to give good resolution over the range of distances. But after this process you still end up with a bunch of x,y points. You could reverse the log process on the x point to give natural values and then supply these to HighCharts to plot on an x-axis. The shape of the plot and line would look identical, the only thing that would change would be the labels on the x-axis. These would show base pairs instead of log10 basepairs.

This is a super minor tweak that almost no-one will notice, so I think it's probably easiest if we drop this anyway and focus on the more important stuff 😄

**Comment by cerebis** [comment]
I have modified the histogram plot somewhat. Trialled going back to raw counts, but as the bins vary width, this leads to a distortion. Instead, I have kept the normalised density and activated the log button. I also found that I could change the label of the "Counts" button, but this was not in the documentation for linegraph.

**Comment by cerebis** [comment]
> This is a super minor tweak that almost no-one will notice, so I think it's probably easiest if we drop this anyway and focus on the more important stuff 😄

You'll have to point out what remains to be address. I am guessing its removing tabular data, but I am concerned we might be miles apart on this issue. I would like the report to remain functional to myself also. Is it possible to have module-specific runtime options with MulitQC?

**Comment by ewels** [comment]
Nice, thanks! I’ll take another look next week (I haven’t looked since you did any changes).

I’m not too worried that we’re miles apart - sorry if my tone sounded strong above. I like to raise the issue of tables when it comes up, but if you have good reasons for the tables to be there then that’s fine to keep them.

You can have module specific config, yes - this is something else that I should probably add to the docs. You use the same `config` object that has the MultiQC configuration in but use `getattr` and `get` defensively so that stuff still works if your module config settings are not there. For example:

```python
collapse_complementary = getattr(config, 'bcftools', {}).get('collapse_complementary_changes', False)
```

Which works with the following config, if set:

```yaml
bcftools:
    collapse_complementary_changes: true
```

Phil

**Comment by cerebis** [comment]
I have decied for the time being it would be more prudent to just simplify things. I have reduced the table columns and remove one entirely.

**Comment by cerebis** [comment]
Hi Phil, could you please take a look at this pull request.

**Comment by ewels** [comment]
Hi @cerebis,

I'm so sorry for the long silence. I've started working part time with paternity leave and it doesn't leave much time for MultiQC.

I think this module looks fantastic, you've done a great job 🙌🏻  I usually ask people to delete all commented-out code, but as it's essentially functional here then I think I will make an exception in case we want to add this back in later on.

This must be one of the prettiest MultiQC modules so far with the colour palette that you've chosen, it looks fab. 🌈 

Phil

**Review by ewels — APPROVED** [review]

**Comment by ewels** [comment]
ps. When running with the |test data](https://github.com/ewels/MultiQC_TestData/tree/master/data/modules/qc3C) I get a lot of the following error messages:

```
qc3C : The entry 'n_low_cov' was not found in the qc3C JSON file './kmer_out/DRR177162/report.qc3C.json'
```

Are these intentional or is the test data now out of date?

Phil

**Comment by ewels** [comment]
Just saw https://github.com/ewels/MultiQC_TestData/pull/164 🤦🏻 👍🏻

**Comment by cerebis** [comment]
Thanks for attending to the PR Phil! 

I appreciate the comments about colour. I do like to try and make this nice looking and for colours to achieve something. I will tip my hat to ColorBrewer and in particular coolors.co.

Glad to hear your gap in attention to MultiQC is due to parental leave and not associated with COVID-19. Congratulations.

**Pull request merged**